[JENKINS-32442][JENKINS-32776] New API Token system

Wadeck opened the pull request:
TL;DR: only the hash is stored now and you can have multiple tokens

### Context and improvements
#### What was the problem
* JENKINS-32442: When a user is created, an API token was generated by default. Potentially the user was not even aware of that.
* JENKINS-32776: Tokens are stored in a recoverable format. So you succeed to find a backup of Jenkins instance with the secret, you could retrieve the plain API Token and impersonate user on the real instance without any warnings of security.
#### Ease of migration
* As long as the users still have Legacy API Token registered on their account the legacy API will still work. If a user does not have any Legacy API Token, the calls will simply returns "deprecated".
* There are two legacy behaviors:
  1. Automatic generation of Legacy API Token when a user is created
  1. Capability to change the Legacy API Token
  * Both of those features can be disabled in the Security Configuration page.

<details>
<summary>Screenshot</summary>

![token_configuration](https://user-images.githubusercontent.com/2662497/35453524-d0048bc0-02ca-11e8-970e-99da64fdef3a.png)

</details>

* As long as they are not disabled, an administrative monitor will remind the administrator to do that. (one for each feature).

<details>
<summary>Screenshot</summary>

![token_warnings](https://user-images.githubusercontent.com/2662497/35453625-3ad097e6-02cb-11e8-8c44-e296dff57b58.png)

</details>

* When a user has a Legacy API Token, we display a warning to ensure he's aware of the problem

<details>
<summary>Screenshot</summary>

![token_legacy](https://user-images.githubusercontent.com/2662497/35453429-74f192dc-02ca-11e8-82a8-7721168d486c.png)

</details>

* The existing unit tests are passing with the current implementation. (just force the legacy behavior in such tests)
* The <code>~~Setup~~UpgradeWizard</code> is adapted. If you upgrade your instance you will keep the legacy behavior by default and if it's a new install, you will have a the new behavior by default.
* An admin monitor will inform about user that sill have legacy token. With ability to revoke in mass. More details in [this comment](https://github.com/jenkinsci/jenkins/pull/3271#issuecomment-375679802).

#### Hash storage / verification
* Storage of new tokens is done using a ~~BCrypt~~ `SHA-256` hash
* Storage of Legacy API Token is both using the current situation and the new hash system, in order to provide statistics on them as well.
* ~~Default configuration of BCrypt uses 2^13 iterations and can be adjusted using a system property (`jenkins.security.ApiTokenStore.bcryptLogRound`).~~
* ~~Goal of the BCrypt hashing is to take between 100ms and 1s to be computed.~~
* ~~Every token has a different salt in order to avoid a potential weakness.~~ As there is a sufficiant entropy, there is no requirement of salting here (i.e. bruteforce is a lot too costly).

<details>
<summary><del>How the performance issue is solved</del></summary>

Corollary of both previous points: if the user has tons of tokens, the performance will be very bad. 
  * To avoid such issue, a `prefix` is used. This prefix is put in plain text before the token. 
  * This way, we can manage to check only a smaller amount of token instead of checking all.
  * It's like if we are using multiple login for a given user since it's plain text but without annoying the user to change the login in function of the application.
  * When we generate a new prefix, we just check to create one that is not used currently
  * Birthday problem: with 75 tokens there are 50% chance to have a collision (over 4096 possibilities). As we try multiple times to avoid collision, we can manage lots of tokens before the first collision.

</details>

#### New features / UI-UX changes
* When a token is found to match the request token, the counter of use of this token is incremented and the last use date updated.
* Users can now observe if they have tokens that are not / rarely / too long time ago used and decide to revoke them
* They can also know when the token was created (or converted from Legacy) in order to apply token rotation policy is any. For the moment just a change in color is applied when your token is older than 180 days (orange) and 360 (red).

<details>
<summary>Screenshot</summary>

![token_color_age](https://user-images.githubusercontent.com/2662497/35453669-5af20320-02cb-11e8-906c-5a2fa8590390.png)

</details>

* You can generate new tokens, rename the tokens after creation (if you have multiple applications using Jenkins tokens), revoke them individually. 

<details>
<summary>Screenshots</summary>

Warning before generation:
![token_generation](https://user-images.githubusercontent.com/2662497/35453720-867adb66-02cb-11e8-9dc0-84c8727e62c3.png)

Token generated:
![token_generated](https://user-images.githubusercontent.com/2662497/35453735-8fef76f2-02cb-11e8-9a26-c08426c3ac0e.png)

Confirm alert in case of revoke:
![token_confirm](https://user-images.githubusercontent.com/2662497/35453789-c5a0e5ce-02cb-11e8-86e2-2effaa263161.png)

</details>

* Those features allow user to apply best practices in terms of credentials management: 
  * Not the same token for every application 
  * Rotation of credentials.

#### Drawbacks
* ~~UI elements used are specific for tokens since nothing (that I am aware of) exist that satisfy the need~~ Now using `<f:repeatable>...`
* Some specific JavaScript is used to provide the desired behavior.

See [JENKINS-32442](https://issues.jenkins-ci.org/browse/JENKINS-32442) and [JENKINS-32776](https://issues.jenkins-ci.org/browse/JENKINS-32776). Both corrections are strongly related, put together to avoid merge conflicts.

### Proposed changelog entries:
* API Token system reworked. Now the user is able to manage multiple API Tokens, generating new, revoke them, use a name to ease the link with usage.
* No breaking change since the legacy features can be disabled progressively:
  * generation of legacy token for new user
  * capability to generate new legacy token even if the user does not have one (unidirectional migration)
* Tokens are now stored only as a hash for the verification
* You can now observe the usage of your tokens with the number of use and the last use date.

### Submitter checklist

- [x] JIRA issue is well described
- [x] Changelog entry appropriate for the audience affected by the change (users or developer, depending on the change). [Examples](https://github.com/jenkins-infra/jenkins.io/blob/master/content/_data/changelogs/weekly.yml)
      * Use the `Internal: ` prefix if the change has no user-visible impact (API, test frameworks, etc.)
- [x] Appropriate autotests or explanation to why this change has no tests

<!-- For new API and extension points: Link to the reference implementation in open-source (or example in Javadoc) -->

### Desired reviewers

@reviewbybees @daniel-beck @recena (esp. for UI)

Wadeck committed:
[JENKINS-32442] Disable API Tokens when user is created
- admin monitor + configuration
- preparation for rewrite API Token

Wadeck committed:
[JENKINS-32776] API Tokens not more stored in a recoverable format
- rewrite the groovy view into jelly one
- remove old config_xx.properties
- add capabilities on token: list/generate/revoke
- stored using BCrypt hashes
- migration from Legacy API Token

Wadeck committed:
- adapt for new UI style on warning message
- correct the help file name

oleg-nenashev commented on core/src/main/resources/jenkins/security/ApiTokenProperty/config.groovy:
```diff
@@ -1,36 +0,0 @@
- * The MIT License
- *
- * Copyright (c) 2011, CloudBees, Inc.
- *
```
Ensure there is no inheriting classes using it somehow.
It would be strange if it happens, but who knows...

Wadeck replied:
By replacing the groovy by a jelly, there is no problem no ?

daniel-beck commented:
@jenkinsci/code-reviewers assemble

stephenc commented:
Does the token hashing mean that every API request will pay a hashing penalty?

Would it be a good idea to populate an in-memory cache with “cheap hashes” (eg salted and then sha256 or concat(sha256,md5) ) that have been verified in the past 5 minutes (make the time limit configurable by system property) so that people using the API token heavily only pay the cost when initiating the “session”?

Wadeck committed:
Improvement to keep pure compatibility with legacy system

Wadeck committed:
- add a short-living cache for the hash with just sha256(salt + plainToken) to speed up things after the first "complete" check

Wadeck committed:
- fix for test

Wadeck commented:
> Does the token hashing mean that every API request will pay a hashing penalty?

It was, before the implementation of your idea using a expiry-cache. @stephenc PTAL at the implementation to know if it suits you / your idea.

oleg-nenashev commented:
I propose to put it on hold so it lands in weeklys after the LTS cutoff. WDYT @Wadeck @daniel-beck ?

Wadeck commented:
@oleg-nenashev I think it will not even ready for the LTS :) so no problem to put on hold if you think it's required. I need to gather some feedbacks on the implementation proposal, then write tests to support the new feature.

daniel-beck commented:
On-holding to prevent merges into the upcoming weekly.

daniel-beck commented:
@Wadeck  To clarify, what's the state of this? Lacking tests before reviews are tentatively approving things?

Wadeck commented:
@daniel-beck exactly, I want to gather feedbacks / reviews before writing tests / adapting to the comments.

bbc2 commented:
For what it's worth, I've had a quick look at the specification and I like the fact that tokens won't be recoverable.

There's something I'm curious about however.  Bcrypt is a good hashing function that compensates the frequent lack of entropy in user-generated passwords by using a salt and a work factor.  On the other hand, if the API tokens from Jenkins are uniformly random and of a sufficient length (like 16 bytes, which would require an average of 2^127 attempts with bruteforce), it seems to me that a good and fast hash function would be suitable.  That would avoid the need for caching hashes in-memory.  I might be missing something about the API token system and its security requirements though.  I hope this makes sense.

dwnusbaum commented on core/src/main/java/jenkins/security/ApiTokenProperty.java:
```diff
@@ -176,29 +317,153 @@ public String getDisplayName() {
+                return HttpResponses.forbidden();
+            }
+            
+            if (StringUtils.isBlank(newTokenName)) {
```
Specific API tokens are not accessible via URL, so we don't need to worry about validating the names any more than this, right?

Wadeck replied:
Exactly, the name is just a way to help user to remember which token is used for what. Nothing depends on the name except the display.

dwnusbaum commented on core/src/main/resources/jenkins/security/ApiTokenProperty/config.jelly:
```diff
@@ -0,0 +1,110 @@
+    
+    <f:entry title="${%CurrentTokens(instance.tokenList.size())}" help="${descriptor.getHelpFile('tokenStore')}">
+        <ul class="token-list">
+            <j:forEach var="token" items="${instance.tokenList}">
```
What happens if multiple tokens in this list have the same name? I don't think it would be an major issue but it might be confusing. Maybe when tokens are renamed or created you could check for duplicates?

Wadeck replied:
I don't think it's necessary since you have the statistics to differenciate them in case you have multiple tokens with the same name.

dwnusbaum reviewed (commented):
Still looking through this, just had some minor comments. I am very positive about the overall idea, nice work!

Wadeck commented on core/src/main/java/jenkins/security/ApiTokenProperty.java:
```diff
@@ -176,29 +317,153 @@ public String getDisplayName() {
+            }
+
+            if (Jenkins.getAuthentication() == ACL.SYSTEM) {
+                // SYSTEM user is always eligible to see tokens
```
From my PoV it's more a defensive approach and to be consistent with `#hasPermissionToSeeToken`. 

If we think the `@Restricted` is sufficient to protect access, I can remove the check in `SYSTEM`

Wadeck commented:
> it seems to me that a good and fast hash function would be suitable

@bbc2 thank you for reviewing this PR, initially I was not directly convinced by your idea since I don't want token to be "recoverable" too easily if someone find a backup or a way to access the hashes. But after reading some articles, making some computations, comparing time used by BCrypt compared to a single pass of SHA-256 (factor between thousand and million) but at the scale of the number of possibilities (10 ^ 37), it's nothing. I am convinced.

I propose to remove the cache mechanism and also replace BCrypt by just HMAC-SHA256 (with new key, not reusing the `API_KEY_SEED` from `ApiTokenProperty`).

If anyone is against that proposal, please submit your arguments.

> `TreeMap` and `LinkedList` proposal from @dwnusbaum 

Modifications done locally, nice idea, good amount of code saved and better readability, just a bit more complex for the storage of information (list vs map). I will not put those modifications online before having feedback on the above proposal.

@daniel-beck @stephenc WDYT about replacing BCrypt ?

jglick commented on core/src/main/java/jenkins/security/ApiTokenPropertyDisabledDefaultAdministrativeMonitor.java:
```diff
@@ -0,0 +1,66 @@
+/**
+ * Monitor that the API Token are not generated by default without the user interaction.
+ *
+ * @since TODO
```
`@since` is meaningless for `@Restricted` members.

jglick commented on core/src/main/java/jenkins/security/ApiTokenStore.java:
```diff
@@ -0,0 +1,634 @@
+    private static final SecureRandom RANDOM = new SecureRandom();
+    
+    private static final Comparator<HashedToken> SORT_BY_LOWERCASED_NAME =
+            Comparator.comparing(hashedToken -> hashedToken.getName().toLowerCase());
```
Do you mean to do a locale-sensitive sort here? If so, use the appropriate methods (`Collator` IIRC). If not, specify `Locale.ENGLISH`.

jglick commented on core/src/main/java/jenkins/security/ApiTokenStore.java:
```diff
@@ -0,0 +1,634 @@
+     */
+    private static final HashCache HASH_CACHE = new HashCache();
+    
+    private List<HashedToken> tokenList;
```
XStream will store any `Comparator`, so do not make such a field non`transient`.

jglick commented on core/src/main/resources/jenkins/security/ApiTokenProperty/config.properties:
```diff
@@ -0,0 +1,10 @@
+CurrentTokens=Current tokens
+TokenCreation=Created {0} day(s) ago
+RenameToken=Save the new name of the token
+LegacyToken=We strongly recommend you to revoke this Legacy API Token as it''s not as secure as the others
```
`it is`

jglick commented on test/src/test/java/jenkins/security/BasicHeaderProcessorTest.java:
```diff
@@ -8,6 +8,7 @@
 import hudson.model.User;
 import hudson.util.HttpResponses;
 import org.junit.Before;
+import org.junit.Ignore;
```
revert

jglick commented on test/src/test/java/jenkins/security/ApiTokenPropertyTest.java:
```diff
@@ -107,6 +109,7 @@ public void adminsShouldBeUnableToSeeTokensByDefault() throws Exception {
     @Issue("SECURITY-200")
     @Test
     public void adminsShouldBeUnableToChangeTokensByDefault() throws Exception {
+        j.jenkins.setCrumbIssuer(null);
```
Why?

jglick commented on test/src/test/java/jenkins/security/ApiTokenPropertyTest.java:
```diff
@@ -130,14 +133,15 @@ public void adminsShouldBeUnableToChangeTokensByDefault() throws Exception {
         
         WebClient wc = j.createWebClient();
-        wc.addRequestHeader("Authorization", "Basic " + Scrambler.scramble(username + ":" + token));
+        wc.addRequestHeader("Authorization", "Basic " + Scrambler.scramble(id + ":" + token));
```
I thought we had something in `JenkinsRule.WebClient` for this?

jglick commented on core/src/main/java/jenkins/security/ApiTokenStore.java:
```diff
@@ -0,0 +1,634 @@
+        // to ease the modification of the token through the UI
+        private transient String uuid;
+        private String name;
+        private Date creationDate;
```
check XStream format; maybe prefer `long`

jglick reviewed (changes requested):
bcrypt is a _password_ hashing mechanism. It is the best-of-breed algorithm for verifying _passwords_ and making sure they are not recoverable. For randomly-generated API _tokens_ none of those considerations apply: brute-force search would be intractable even if the hash check took a single clock cycle, and no salt is needed since no rainbow table attacks would make sense. So I see no reason to make things complicated by using bcrypt.

You do not need `HMACConfidentialKey` (or MACs generally) in `ApiTokenStore` either. That is only useful for today’s `ApiTokenProperty.API_KEY_SEED` because it supports a _default_ token hashed out of the user ID (and `master.key`). You can directly store a SHA-256 (or whatever) in `HashValue.hash`.

Wadeck committed:
- devin's proposal

Wadeck commented on core/src/main/java/jenkins/security/ApiTokenProperty.java:
```diff
@@ -138,36 +195,120 @@ private boolean hasPermissionToSeeToken() {
+        // just to keep the same level of security
         user.checkPermission(Jenkins.ADMINISTER);
+
+        LOGGER.log(Level.WARNING, "Deprecated usage of changeApiToken");
```
I think that could help developer of plugin that are using directly the method to have a warning instead of silently doing / not doing things.

If you think it's useless, I will remove it.

Wadeck commented on core/src/main/java/jenkins/install/SetupWizard.java:
```diff
@@ -122,6 +123,10 @@ public SetupWizard() {
     
+                    // Disable the legacy system of API Token
+                    ApiTokenPropertyConfiguration.get().setCreationOfLegacyTokenDisabled(true);
+                    ApiTokenPropertyConfiguration.get().setTokenGenerationOnCreationDisabled(true);
```
Do you have another possibility to avoid a fresh install to show warnings from monitor even after a run of the wizard because they are enabled by default ? (the change from negative to positive approach does not impact this behavior).

IIUC one of the wizard's goal is to provide "good" defaults.

daniel-beck replied:
Would be better to disable legacy by default and figure out whether it's an upgrade from earlier, and only enable it in that case.

Not sure what the best way is, @dwnusbaum and @jglick might have ideas as they've recently worked on initialization IIUC.

(And in general it's probably useful for some of the existing options to be moved out of the wizard. Perhaps we can keep it here but plan to have a followup change that changes the default options and instead detects an update?)

dwnusbaum replied:
Regarding enabling on upgrade, I think you could put the logic in `UpgradeWizard.java#initializeState`, which corresponds to `InstallState.UPGRADE`, but it would need to be tested.

Wadeck replied:
@dwnusbaum thank you for the tip, I will test that :)

Wadeck replied:
I tested the existence of the config file of `ApiTokenPropertyConfiguration` during the upgrade process to know if I must apply default or not.

Wadeck committed:
- adjust Devin's proposal
- address Jesse's review first part

Wadeck committed:
Change the approach for the configuration, when upgrading, we force the default secured version if you did not have ApiTokenPropertyConfiguration configured before.
- inverse the boolean after Jesse's comment

Wadeck commented:
Changes: 
- BCrypt replaced by simple SHA-256 without salt / MAC.
- Configuration using enabled instead of disabled + upgrade process instead of new install process

olivergondza committed:
Add tests for REST API authenticated by token in parameter (#2768)

* Add tests for REST API authenticated by token

* Avoid adding the crumb that is not needed anymore

(cherry picked from commit b652b07)

Wadeck committed:
- correction of two bugs, when the user does not have any token, the legacy generation is not displayed and when we load config with user with legacy token
- de-final-ize the properties to be used in Groovy Script

Wadeck committed:
- remove useless code

Wadeck committed:
Merge branch 'master' into JENKINS-32442-32776_HASHED_TOKEN

Wadeck committed:
provide the token id when we generate a new token, could be used to automate the generation / revocation of token

Wadeck committed:
Merge remote-tracking branch 'origin/JENKINS-32442-32776_HASHED_TOKEN' into JENKINS-32442-32776_HASHED_TOKEN

jglick committed:
TextFileTest.tail failed on Windows when using a source checkout with Unix newlines.

(cherry picked from commit 3a81319)

dwnusbaum commented on core/src/main/resources/jenkins/security/ApiTokenProperty/config.jelly:
```diff
@@ -0,0 +1,109 @@
+    </f:entry>
+
+    <j:if test="${descriptor.hasCurrentUserRightToGenerateNewToken(it)}">
+    <!--<j:if test="${instance == null || instance.hasCurrentUserPermissionToGenerateTokenForThisUser()}">-->
```
The commented out code can be removed.

dwnusbaum commented on core/src/main/resources/jenkins/security/ApiTokenProperty/help-tokenStore.html:
```diff
@@ -0,0 +1,12 @@
+<div>
+    The API tokens can be used for authenticating yourself in the REST API call.
+    See <a href="https://jenkins.io/redirect/api-token">our wiki</a> for more details.<br/>
+    The associated username with the tokens is the regular one of your account.<br/>
```
I would rephrase this to: The username associated with each token is your Jenkins username.

dwnusbaum commented on core/src/main/resources/jenkins/security/ApiTokenProperty/help-tokenStore.html:
```diff
@@ -0,0 +1,12 @@
+    <br/>
+    The best practices in terms of tokens are:
+    <ul>
+        <li>Use a different token for each application (if one application is compromised, you do not loose data from other)</li>
```
loose -> lose

dwnusbaum commented on core/src/main/resources/jenkins/security/ApiTokenProperty/help-tokenStore.html:
```diff
@@ -0,0 +1,12 @@
+    See <a href="https://jenkins.io/redirect/api-token">our wiki</a> for more details.<br/>
+    The associated username with the tokens is the regular one of your account.<br/>
+    <br/>
+    The best practices in terms of tokens are:
```
Maybe rephrase: Some good practices for keeping your API tokens secure are:

dwnusbaum commented on core/src/main/resources/jenkins/security/ApiTokenProperty/help-tokenStore.html:
```diff
@@ -0,0 +1,12 @@
+    The best practices in terms of tokens are:
+    <ul>
+        <li>Use a different token for each application (if one application is compromised, you do not loose data from other)</li>
+        <li>Regenerate the tokens every x weeks/months. We put an indicator concerning the age of the token.</li>
```
Maybe mention a specific timeframe (6 months?) instead of just using `x`.

Wadeck replied:
Also added `(depending on your context)`

dwnusbaum reviewed (commented):
Just some minor comments

Wadeck committed:
- remove modification in SetupWizard

Wadeck commented on core/src/main/java/jenkins/security/ApiTokenProperty.java:
```diff
@@ -61,49 +73,90 @@
-     * If enabled, shows API tokens to users with {@link Jenkins#ADMINISTER) permissions.
-     * Disabled by default due to the security reasons.
+     * If enabled, the users with {@link Jenkins#ADMINISTER} permissions can generate new legacy tokens for
+     * other users. Normally only a user can generate tokens for himself.<p>
```
Reworded completely as it was no more accurate

Wadeck committed:
remove the prefix as it is not more useful (was for BCrypt but no more with SHA-256)

Wadeck committed:
adjust some whitespaces and removal of since

dwnusbaum commented on core/src/main/java/jenkins/security/ApiTokenStore.java:
```diff
@@ -218,89 +178,41 @@ private void addLegacyToken(@Nonnull Secret legacyToken) {
-     */
-    private String getPrefixOfToken(String token) {
-        return token.substring(1, 4);
+        return token.substring(0, 1);
```
`endIndex` is exclusive, so this is just the first character. You want `substring(0, 2)`.

dwnusbaum reviewed (commented):
I think there is a small issue with `getVersionOfToken`. You can merge with master to fix `hudson.util.TextFileTest`, but I am not sure about the other test failures.

Wadeck committed:
correct exclusive boundary mistake

Wadeck committed:
correction of some tests that were using the withApiToken without legacy behavior

Wadeck committed:
- correct last failing test, required legacy behavior to work

Wadeck commented:
@dwnusbaum thank you for the note, tests are now passing!

daniel-beck commented on core/src/main/java/jenkins/security/ApiTokenProperty.java:
```diff
@@ -61,49 +73,90 @@
+    @SuppressFBWarnings("NP_NONNULL_RETURN_VIOLATION")
     public String getApiToken() {
-        return hasPermissionToSeeToken() ? getApiTokenInsecure() 
+        LOGGER.log(Level.WARNING, "Deprecated usage of getApiToken");
```
This should probably log the stack trace (perhaps separately on FINE). Otherwise this information is useless.

It's also called when viewing the configure page of a user with a legacy token so I'm not sure a warning is appropriate here.

daniel-beck commented on core/src/main/java/jenkins/security/ApiTokenProperty.java:
```diff
@@ -176,29 +317,169 @@ public String getDisplayName() {
+            // you are the user or you have ADMINISTER permission
+            u.checkPermission(Jenkins.ADMINISTER);
+
+            LOGGER.log(Level.WARNING, "Deprecated action /changeToken used, consider using /generateNewToken instead");
```
This also happens when enabling legacy token options in the security configuration, and seems therefore inappropriate for a warning.

daniel-beck commented:
I tested the PR build 14.

First, this is a fantastic change 👍 I'm really looking forward to see this land in Jenkins! 🎉 

I've discovered two not-quite-minor issues in testing:
* There should be an admin monitor, listing all users with legacy API tokens, and when their legacy tokens were last used, that offers an option to bulk revoke them all (or only those without recorded usages, so that it can be done a week after updating, and retains all tokens that were used since update). Some instances have dozens or hundreds of users, revoking all the tokens one by one will be very cumbersome.
* Updating from 2.107 triggers an Old Data Monitor Warning with the message "`@since TODO`" but it's not clear why. Re-saving users will not remove legacy tokens, so that's not it.

Some relatively minor things that still should be changed:

* Do we actually want to retain the ability to create legacy tokens? I'm not sure these are valid UI options. Sure, retain the existing tokens, but creating new ones? Do they have any value?
  * The global configuration options lack help text. It is not obvious what they do, or why they exist in the first place.
* We should change some doc wording from "unsafe" to "stored on disk in a recoverable format" or similar. Legacy tokens aren't terrible, they're just recoverable credentials.
* The watermark for the token name ("Default name") seems to indicate a name will be generated, but it rejects empty names. Use a more explicit watermark. Possibly allow name generation if unspecified ("Api Token created on $date $time" perhaps?).
* When I have a user without legacy token and 'Show Legacy Token', it will show 'deprecated' rather than e.g. 'This user currently does not have a legacy token'.

Additional notes that I don't think need fixing unless very simple to implement. Noted here for your and other reviewers' consideration:

* Token creation triggers the form dirty flag, so on navigating away, it asks about unsaved changes that don't exist. Not worse than the legacy token change, at least (i.e. weird but not a regression).
* Have a token on Configure page load. Create a new one so it shows. Two results:
  * The button to create a new token disappears (not clear why I cannot create another token).
  * When you now delete the pre-existing token, it will show "no tokens".
* Revoking legacy tokens has weird UI
  * I can revoke the legacy token, and then click 'Show Legacy Token', and it will show me the revoked token value.
  * I can revoke the legacy token of another user, and 'Show Legacy Token' is still visible until I reload the form.

daniel-beck commented:
I'm not convinced storing API tokens in the user config.xml is a great idea -- at least not the use counter.

daniel-beck commented:
> No breaking change since the legacy features can be disabled progressively:
> 
> * generation of legacy token for new user
> * capability to generate new legacy token even if the user does not have one (unidirectional migration)

Thinking about this some more, neither is necessary for this to not be a breaking change: New users by definition haven't used their /API tokens in the past, so aren't affected by a change to this system. Existing users' tokens are retained, so existing scripts will just continue to work.

I even doubt these options even need to be system properties. with migration tests to ensure existing tokens continue to work, we should be good.

Wadeck committed:
Address using Daniel's comments (first part)
- improve UI/UX using repeat instead of advanced
- placeholder "Default Token" is transformed into a real default "unique" name after generation and can be renamed
- remove trigger of Old Data Monitor (custom monitor not yet started)
- log (finer) the stack trace in case of deprecated call to getApiToken()
- reduce warning to fine for deprecated usage
- change the "deprecated" token value to "This user currently does not have a legacy token"
- in case we revoke the legacy token from the store, the plain legacy token is now shown as "Legacy token was revoked"
- improve confirm.js to allow developer to specify part of form that are not expecting confirm behavior

daniel-beck commented on war/src/main/webapp/scripts/hudson-behavior.js:
```diff
@@ -2515,6 +2515,9 @@ function buildFormTree(form) {
                 jsonElement = e;
                 continue;
             }
+            if(e.hasClassName('json-ignore')){
```
Is there really no other element to support content that should be ignores in JSON?

Wadeck replied:
I just changed that after trying the write a first comment. This code only remove the JSON part but actually the token value is submitted. So I changed the input(type=hidden) to a simple span with style and remove this part.

daniel-beck commented on core/src/main/resources/lib/form/confirm.js:
```diff
@@ -15,6 +15,20 @@
+      }
+      // to allow sub-section of the form to ignore confirm
+      // especially useful for "pure" JavaScript area
+      var ignorePanel = element.up('.ignore-confirm-panel');
```
I wonder whether a more general name would be preferable here, as this effectively becomes API. Confirmation panel is a specific effect of a 'dirty' or modified form.

Wadeck replied:
`no-dirty-trigger` ?
`stay-pristine` ? (mix of Angular form validation and Britney Spears) _(joke)_

Wadeck committed:
avoid submitting the plain token value if the user submit the form when there is a new token freshly saved
- also remove the "json-ignore" which is no more required with this new approach

Wadeck commented:
First part of correction:
- I added a feature in `confirm.js`:
  1. `ignore-confirm` a button / an input with such class will not trigger the confirm behavior
  2. `ignore-confirm-panel` same as before but it just has to be present in its ancestors (useful for a whole panel that is managed essentially by JavaScript.

<details><summary>Revoke legacy token behavior (gif inside)</summary>

![legacy-revoked](https://user-images.githubusercontent.com/2662497/37710762-ec936306-2d0e-11e8-8b64-9a4b8686b42f.gif)

</details>

<details><summary>Empty list behavior (gif inside)</summary>

![empty-list](https://user-images.githubusercontent.com/2662497/37710770-f526c53a-2d0e-11e8-8c9e-e9b825df531d.gif)

</details>

Wadeck committed:
change from javascript &nbsp; to the correct jelly tag thx to Daniel

Wadeck committed:
add migration test using localData
- also add test on the generate api call

Wadeck committed:
- change package (better hierarchy)
- correct small text when instance is null
- add monitor

Wadeck commented:
New monitor to revoke the legacy token.


<details><summary>Tooltip (gif)</summary>

![legacy-monitor-tooltip](https://user-images.githubusercontent.com/2662497/37832123-f6b40a26-2ea7-11e8-8d37-fc3d6bef68ea.gif)


</details>

<details><summary>Selection feature (gif)</summary>

![legacy-monitor-select](https://user-images.githubusercontent.com/2662497/37832120-f4c85406-2ea7-11e8-9ad0-3481ebf849e4.gif)


</details>

<details><summary>Revocation (gif)</summary>

![legacy-monitor-revoke](https://user-images.githubusercontent.com/2662497/37832117-f2dcf4c6-2ea7-11e8-987b-3989b7c1856c.gif)


</details>

This monitor is triggered when at least one user still have a legacy token.

For the moment the check is done on each users, every time the monitor is checked. I dunno if it could cause performance issue on large instance. If I understand correctly the code, the user list and the `ApiTokenProperty` are loaded in memory. WDYT ?

Wadeck committed:
- add unit test for the monitoring page (include filter work)
- correct placement of stapler:adjunct to avoid putting <link> before the html tag
- decrease the javascript version required to run resources.js (for unit test)
- re-move the zip file due the package migration that was done on it but was not required

Wadeck committed:
ApiTokenStore is now saved / loaded from a file aside config.xml in user folder
- this allows the stats to be incremented without changing the config.xml of the user

Wadeck commented:
The `ApiTokenStore` is now saved separately from the `config.xml` of the user.

rsandell commented:
With the introduction of Jenkins Configuration as Code is it still sensible to use the _SetupWizard_ to set security defaults? Because that would mean IIUC that the config code would need to explicitly set those values and that doesn't sound safe to me since the setupwizard I guess would mostly be turned off when using conf as code.

daniel-beck commented:
@rsandell It's not like this PR introduces a new problem -- other defaults are also crazy when bypassing the wizard, and since CasC isn't even out yet (alpha doesn't count) I'm not worrying too much about it yet.

Wadeck commented:
@rsandell initially it was using the SetupWizard to set the correct configuration. But now after the comment from Jesse, the default value is the one we expect and in case of upgrade, we configure the legacy behavior (changed from default).

So if you skip the Wizard for a new install, you have the "secured" configuration.

Wadeck committed:
Change the name of the classes for ignore the confirm popup

Wadeck committed:
Separate the statistics from the token store
- different file for better versioning / synchronization
- ability using a configuration to avoid completely the statistics
- add unit / integration tests
- adjust the configuration page

Wadeck committed:
Merge branch 'master' into JENKINS-32442-32776_HASHED_TOKEN

Wadeck commented:
Token storage and token usage are now separated.
- storage is put back into `users/<login>/config.xml`
- and statistics into `users/<login>/apiTokenStats.xml`
- add multiple tests
- statistics can be disabled in the API Token configuration section

jvz commented on core/src/main/java/jenkins/security/apitoken/ApiTokenStats.java:
```diff
@@ -0,0 +1,262 @@
+                .findFirst()
+                .orElse(null);
+        
+        if (stats == null) {
```
I feel like using `Optional` and `null` at the same time is a bit of an anti-pattern.

Wadeck replied:
I changed the code, it looks like better, 👍

jvz commented on core/src/main/java/jenkins/security/apitoken/ApiTokenStore.java:
```diff
@@ -0,0 +1,371 @@
+
+public class ApiTokenStore {
+    private static final Logger LOGGER = Logger.getLogger(ApiTokenStore.class.getName());
+    private static final SecureRandom RANDOM = new SecureRandom();
```
In Java 8, it may be preferable to use `SecureRandom.getInstanceStrong()` to prevent any bad SecureRandom implementations by being used.

Wadeck replied:
Concerning this point I was initially positively surprised to see a new method for that, but then when I use it, it throws a check exception, so the code becomes a lot longer (to fallback to default version in case of exception) and at that point I tried to investigate a bit the topic.

Here are different links ([issue tracker of Akka](https://github.com/akka/akka/issues/20908), [blog post](https://tersesystems.com/blog/2015/12/17/the-right-way-to-use-securerandom/), [stackoverflow](https://stackoverflow.com/a/27638413/2567174), [Synopsys blog post](https://www.synopsys.com/blogs/software-security/proper-use-of-javas-securerandom/)) that advice to not use this method due the blocking access of `/dev/random` without a positive impact on security. As Jenkins is often running inside container, the entropy of the system could not large enough.

So I let it like that for the moment except if you have arguments to change it.

jvz replied:
Yikes, I didn't realize it threw a checked exception! Sigh, the things you miss when you're in Scala or Kotlin land. :)

As for blocking on `/dev/random`, from what I've read, that hasn't been an actual problem in years. IIRC, Linux seeds this on boot which is like the only time it blocks, and then afterwards it's a PRNG just like `/dev/urandom`. Not sure how it's implemented in Windows, though, but I'm guessing it's similar.

Wadeck committed:
Better usage of Optional

Wadeck committed:
Merge remote-tracking branch 'origin/JENKINS-32442-32776_HASHED_TOKEN' into JENKINS-32442-32776_HASHED_TOKEN

dwnusbaum commented:
I will need a clear head and some uninterrupted time to actually review the code, but I tried out the feature from a usage only perspective today. I tested a new installation and an upgrade of an installations with existing legacy tokens and did not find any issues actually using either kind of token via the CLI or REST API, which is great!

Here are a few comments from my tests:
* Using a token with the CLI (no specified command) increments the number of token usages by 3. Is this _easily_ fixable? REST API calls appear to work correctly and only increment the number of usages by 1.
* The page that allows an admin to revoke legacy API tokens feels like maybe it should be extended to work with new API tokens and should have a permanent link in /manage (or maybe be part of the manage Users page (/securityRealm)) so that admins can get to it even if there is not currently a warning.
* Some of the data in the legacy token admin monitor is missing for me, but it seems like it shouldn't be empty based on the user configuration page. I don't see any exceptions in the logs: 
<img width="1066" alt="screen shot 2018-05-03 at 17 20 47" src="https://user-images.githubusercontent.com/1068968/39603468-85698fca-4ef6-11e8-959e-f41d13c2a8cb.png">
<img width="898" alt="screen shot 2018-05-03 at 17 22 24" src="https://user-images.githubusercontent.com/1068968/39603483-942cade4-4ef6-11e8-9a57-9d3e22d26ad1.png">

Wadeck committed:
Address Devin's comment
- adjust the way the stats are used for the monitoring page
- correct the re-creation of legacy token on restart

Wadeck commented:
> Is this easily fixable?

@dwnusbaum unfortunately nope. The CLI does first a GET on `/jenkins`, then a GET on `/jenkins/crumbIssuer`, finally a POST on the `/jenkins/cli` (actually twice the POST I dunno why), for a total of 4 usages for a single command. (Using -auth)

> it should be extended to work with new API tokens

Not in the scope of this PR (that should not have been that large ^^) but definitely interesting. I think addition of metadata on the token could be nice, like reducing the scope in terms of authorization, or the validity duration (expiration), etc. In such situation the permanent page could be better but as currently it's only showing the legacy token, it's temporary existence seems fine.

> Some of the data in the legacy token admin monitor is missing for me

Good catch, thank you. This was due to the last migration of the statistics in another file. Corrected now!

daniel-beck commented:
> actually twice the POST I dunno why

IIRC this is to get two streams, one for input, one for putput. There's a query parameter that controls which side it is.

daniel-beck commented:
> then a GET on `/jenkins/crumbIssuer`

Probably obsolete with the recent improvement to not need crumbs for HTTP Basic; should be an independent futute enhancement though.

Wadeck commented:
> not need crumbs for HTTP Basic

@daniel-beck to be precise, it's only true for API token cf. [PR#3129](https://github.com/jenkinsci/jenkins/pull/3129/files#diff-051e63be5278e65ddcf1a8ef98c33ac8R47).

I thought about that when trying to reproduce but actually in the CLI, we do not know if the current "password" provided is really a password or an API Token. Possible solution is to provide a new `--token` and `--tokenFile` parameters like `--password` and `--passwordFile`.

dwnusbaum committed:
Remove unnecessary dash

(Wadeck and I want to see how the on-GitHub PR editing features work in case they are useful for proposing grammar changes)

dwnusbaum commented:
Some other things regarding the UX of creating a token:

<img width="845" alt="screen shot 2018-05-04 at 09 48 51" src="https://user-images.githubusercontent.com/1068968/39631557-773f0c38-4f80-11e8-9c61-c23dd4d53777.png">

* I think it might be better to only display the warning once the new token has been generated
* I would reword the warning to be more direct and less technical like: "Copy this token now, because it cannot be recovered in the future.". 
* If it is easy, I would also change "save" to "create" or "generate" if the token is being created for the first time, but would leave it as "save" when the name is being changed.

Wadeck committed:
Change the warning message

Wadeck committed:
Adjust UI after discussion with Devin
- trigger the confirm message if someone change an existing token name (for the rename) and does not click on Save / Apply
- improve the ignore/force dirty mechanism
- display the warning message about the temporary aspect of the token plain text only *after* the generation
- change "Save" to "Generate" for the new token

Wadeck committed:
uppercase the token

jglick commented:
> in the CLI, we do not know if the provided "password" is really a password or an API Token

I guess you could just check whether it is a hex string of the expected length. If not, it is definitely not an API token. If so, it _could_ be a (randomly-generated) password, but seems unlikely.

jtnord commented on core/src/main/java/jenkins/security/apitoken/ApiTokenStore.java:
```diff
@@ -0,0 +1,381 @@
+     * [2: version][32: real token]
+     * ^^^^^^^^^^^^----------------
+     */
+    private @Nonnull String getVersionOfToken(@Nonnull String token) {
```
unused???

jtnord commented on core/src/main/java/jenkins/security/apitoken/ApiTokenStore.java:
```diff
@@ -0,0 +1,381 @@
+        return null;
+    }
+    
+    public synchronized @CheckForNull HashedToken revokeToken(@Nonnull String tokenUuid) {
```
in general would like to see much more javadoc.
why might this return null for example (without having to read the code)?

jtnord commented on core/src/main/java/jenkins/security/apitoken/ApiTokenStore.java:
```diff
@@ -0,0 +1,381 @@
+    }
+    
+    /**
+     * [2: version][32: real token]
```
this should be a code comment not a javadoc comment.

jtnord commented on core/src/main/java/jenkins/security/apitoken/ApiTokenStore.java:
```diff
@@ -0,0 +1,381 @@
+     */
+    @Immutable
+    private static class HashValue {
+        /** To ease future implementation */
```
useless from a javadoc POV.

jtnord commented on core/src/test/java/jenkins/security/apitoken/ApiTokenStatsTest.java:
```diff
@@ -0,0 +1,224 @@
+            
+            ApiTokenStats.SingleTokenStats stats = tokenStats.updateUsageForId(ID_1);
+            assertEquals(2, stats.getUseCounter());
+            assertTrue(lastUsage.before(stats.getLastUseDate()));
```
NIT: never use assertTrue or false for comparasions that can be better represented as `assertThat(target, lessThan(comparison))`
WIthout the code as is should this fail you have lost information as to the lastUsage and the stats. with the latter if it fails you are shown the two values

jtnord commented on test/src/test/java/hudson/model/JobTest.java:
```diff
@@ -210,6 +211,9 @@ public String getDisplayName() {
     @LocalData
     @Test public void configDotXmlPermission() throws Exception {
+        // legacy behavior re-enabled (could be changed when the webclient will be adapted
+        ApiTokenPropertyConfiguration.get().setTokenGenerationOnCreationEnabled(true);
```
NIT: could be better as a test helper method that is called, so that when the TODO is done it will be easier to change all the test code.

Wadeck committed:
Address James' comment
- also rename TokenId to TokenUuid everywhere

Wadeck committed:
Correct typo

dwnusbaum commented on core/src/main/java/jenkins/security/apitoken/ApiTokenStats.java:
```diff
@@ -0,0 +1,257 @@
+        
+        private static Comparator<SingleTokenStats> COMP_BY_LAST_USE_THEN_COUNTER =
+                Comparator.comparing(SingleTokenStats::getLastUseDate, Comparator.nullsLast(Comparator.naturalOrder()))
+                        .thenComparing(SingleTokenStats::getUseCounter, Comparator.reverseOrder());
```
I don't think you want `Comparator.nullsLast` or `Comparator.reverseOrder()`. My expectation based on `keepLastUpdatedDuplicated` is that you want the comparator to sort the following as A < B < C < D (newest and most used tokens should come last so they overwrite tokens with duplicate uuids in the map), but as it is you get C < B < D < A.
Token A: 0 Uses, Last on `null`
Token B: 2 Uses, Last on 2018-01-05 00:00:00
Token C: 3 Uses, Last on 2018-01-05 00:00:00
Token D: 2 Uses, Last on 2018-02-05 00:00:00

The following comparator might be better:

    Comparator.comparing(SingleTokenStats::getLastUseDate, Comparator.nullsFirst(Comparator.naturalOrder()))
            .thenComparing(SingleTokenStats::getUseCounter);

Wadeck replied:
Yeah very good point (even if we are in corner case in reality), I checked only B < D in my unit test. 👍 

I added more unit test to cover this part of the comparison.

Wadeck commented on core/src/main/java/jenkins/security/apitoken/ApiTokenStats.java:
```diff
@@ -0,0 +1,257 @@
+        }
+    }
+    
+    private void keepLastUpdatedDuplicated() {
```
Yeah good point, I think initially I wanted to name it "keepLastUpdatedFromDuplicated" or something like that, your name is better.

Wadeck committed:
Address Devin's comment
- change comparator null and counter parts
- add unit test to cover this part of the comparison

dwnusbaum commented on core/src/main/resources/jenkins/security/apitoken/ApiTokenPropertyConfiguration/help-creationOfLegacyTokenEnabled.html:
```diff
@@ -0,0 +1,5 @@
+<div>
+    Even if this is disabled (recommended for security reasons), the users are still able to generate new tokens
```
nit: I would reword this blurb to: "This option allows users to generate a legacy API token if they do not already have one. Because legacy tokens are \<strong\>deprecated\</strong\>, we recommend disabling it and having users instead generate new API tokens from the user configuration page."

dwnusbaum commented on core/src/main/resources/jenkins/security/apitoken/ApiTokenPropertyConfiguration/help-tokenGenerationOnCreationEnabled.html:
```diff
@@ -0,0 +1,5 @@
+<div>
+    If this capability is disabled (recommended for security reasons), the users will be forced to use the new system of API token. <br/>
```
nit: I would reword this blurb to: "This option causes a legacy API token to be generated automatically for newly created users. Because legacy tokens are \<strong\>deprecated\</strong\>, we recommend disabling it and having users instead generate new API tokens from the user configuration page as needed."

dwnusbaum commented on core/src/main/resources/jenkins/security/apitoken/ApiTokenPropertyConfiguration/help-usageStatisticsEnabled.html:
```diff
@@ -0,0 +1,7 @@
+<div>
+    When a user is authenticated by an API Token, its usage counter is incremented and its last usage date updated.
```
nit: I would reword this blurb to "If this option is enabled, then the date of the most recent use of each API token and the total number of times it has been used are stored in Jenkins. This allows users to see if they have unused or outdated API tokens which should be revoked."

Maybe we should also add a disclaimer like "This data is stored locally in Jenkins and will not be used for any other purpose" in case someone sees "usage statistics" and gets worried?

dwnusbaum commented on core/src/main/resources/jenkins/security/apitoken/ApiTokenPropertyDisabledDefaultAdministrativeMonitor/message.jelly:
```diff
@@ -0,0 +1,33 @@
+<j:jelly xmlns:j="jelly:core" xmlns:f="/lib/form">
+    <div class="alert alert-warning">
+        <form method="post" action="${rootURL}/${it.url}/act" name="${it.id}">
+            <f:submit name="yes" value="${%Disable legacy generation by default}"/>
```
nit: I would reword this to "Disable automatic generation of legacy API tokens"

dwnusbaum commented on core/src/main/resources/jenkins/security/apitoken/ApiTokenPropertyDisabledDefaultAdministrativeMonitor/message.properties:
```diff
@@ -0,0 +1,2 @@
+warningMessage=A Legacy API Token is currently generated when a new user is created. <br />\
```
nit: I would reword this to "A legacy API token will be generated automatically for newly created users.\<br/\>This behavior is supported for backwards compatibility, but legacy API tokens are deprecated and are not recommend for long-term use. Users should instead generate API tokens as needed from the user configuration page."

Wadeck replied:
correction applied with fixed typo: recommend**ed**

dwnusbaum commented on core/src/main/resources/jenkins/security/apitoken/ApiTokenPropertyEnabledNewLegacyAdministrativeMonitor/message.jelly:
```diff
@@ -0,0 +1,33 @@
+<j:jelly xmlns:j="jelly:core" xmlns:f="/lib/form">
+    <div class="alert alert-warning">
+        <form method="post" action="${rootURL}/${it.url}/act" name="${it.id}">
+            <f:submit name="yes" value="${%Disable that capability}"/>
```
nit: I would reword this to "Prevent users from manually creating legacy API tokens"

dwnusbaum commented on core/src/main/resources/jenkins/security/apitoken/ApiTokenPropertyEnabledNewLegacyAdministrativeMonitor/message.properties:
```diff
@@ -0,0 +1,2 @@
+warningMessage=Users without Legacy API Token have the capability to create new Legacy API Token. <br />\
```
nit: I would reword this to "Users without legacy API tokens are able to generate a new legacy token.\<br/\>This behavior is supported for backwards compatibility, but legacy API tokens are deprecated and are not recommend for long-term use. Users should instead generate API tokens as needed from the user configuration page."

Wadeck replied:
correction applied with fixed typo: recommend**ed**

dwnusbaum commented on core/src/main/resources/jenkins/security/apitoken/LegacyApiTokenAdministrativeMonitor/manage.properties:
```diff
@@ -0,0 +1,33 @@
+title=Manage Legacy API Token usage
+contextMessage=The following users are using Legacy API Token. \
```
nit: I would reword this to: "The following users have a legacy API token. Because legacy tokens are stored in a recoverable format, we recommend migrating to the new API token system. Additionally, in previous version of Jenkins, API tokens were automatically created for every new user. Often, users did not use the token, resulting in a larger attack surface than necessary. Automatic token generation must now be enabled explicitly."

dwnusbaum commented on core/src/main/resources/jenkins/security/apitoken/LegacyApiTokenAdministrativeMonitor/manage.properties:
```diff
@@ -0,0 +1,33 @@
+recommendationMessage=For such reasons, we recommend to: \
+<ul>\
+  <li>revoke the tokens that were never used and</li>\
+  <li>ask users that are effectively using their legacy token to generate a new token using the new system and revoking the legacy one</li>\
```
nit: I would reword this to "ask users that are currently using their legacy token to generate a token using the new system and revoke their legacy token"

dwnusbaum commented on core/src/main/resources/jenkins/security/apitoken/LegacyApiTokenAdministrativeMonitor/manage.properties:
```diff
@@ -0,0 +1,33 @@
+NumOfUse=# of use
+NumDaysSinceLastUse=Days since last use
+HasFreshToken=Fresh token?
+HasFreshToken_tooltip=A fresh token is a token from the new system \
```
nit: I might reword this to "A fresh token is one that was created using the new system after the most recent use of the user's legacy token. Note: the creation date of legacy tokens which have never been used are reset every time Jenkins is restarted, which means that the date may be inaccurate."

dwnusbaum commented on core/src/main/resources/jenkins/security/apitoken/LegacyApiTokenAdministrativeMonitor/manage.properties:
```diff
@@ -0,0 +1,33 @@
+Also it is important to notice that legacy token's creation date is reset \n\
+every time the instance is restarted if they were never saved before. So it was never used, the date is not really accurate.
+HasMoreRecentlyUsedToken=Recent token?
+HasMoreRecentlyUsedToken_tooltip=A recently used token is a token from the new system \
```
nit: I might reword this to "A recently used token is one that was created using the new system and used more recently than the user's legacy token. Note: the creation date of legacy tokens which have never been used are reset every time Jenkins is restarted, which means that the date may be inaccurate."

dwnusbaum commented on core/src/main/resources/jenkins/security/apitoken/LegacyApiTokenAdministrativeMonitor/manage.properties:
```diff
@@ -0,0 +1,33 @@
+that was used the last time after the last use of the legacy token. \n\n\
+Also it is important to notice that legacy token's creation date is reset \n\
+every time the instance is restarted if they were never saved before.
+NoImpactedUser=There are no users with legacy token
```
"with legacy token" -> "with a legacy token"

dwnusbaum commented on core/src/main/resources/jenkins/security/apitoken/LegacyApiTokenAdministrativeMonitor/manage.properties:
```diff
@@ -0,0 +1,33 @@
+NoLastUse=There is no last use date for that token
+RevokeAllSelected=Revoke the selected token(s)
+RevokeAllSelected_confirm=Are you sure about revoking all %num% selected token(s)
+RevokeAllSelected_nothing=No token is selected, please select at least one to do the revocation
```
"select at least one to do the revocation" -> "select at least one to revoke"

dwnusbaum commented on core/src/main/resources/jenkins/security/apitoken/LegacyApiTokenAdministrativeMonitor/message.properties:
```diff
@@ -0,0 +1,3 @@
+warningMessage=There are users that are still using Legacy API Token. \
```
"using Legacy API Token" -> "using a legacy API token"

dwnusbaum commented on core/src/main/resources/jenkins/security/apitoken/LegacyApiTokenAdministrativeMonitor/message.properties:
```diff
@@ -0,0 +1,3 @@
+warningMessage=There are users that are still using Legacy API Token. \
+  That system is not as secured as the new one because it stores the token in a recoverable manner on the disk. <br />\
```
"secured" -> "secure"

dwnusbaum reviewed (approved):
Looks great to me, I just proposed a bunch of changes to user-visible text to try to make things flow better and be more consistent.

Wadeck committed:
Address Devin's rewording points

Wadeck commented on core/src/main/resources/jenkins/security/Messages.properties:
```diff
@@ -24,5 +24,9 @@ ApiTokenProperty.DisplayName=API Token
 ApiTokenProperty.ChangeToken.Success=<div>Updated. See the new token in the field above</div>
 ApiTokenProperty.ChangeToken.SuccessHidden=<div>Updated. You need to login as the user to see the token</div>
+ApiTokenProperty.ChangeToken.CapabilityNotAllowed=<div>Capability to generate a legacy token without an existing one is currently disabled in the security configuration</div>
+ApiTokenProperty.NewTokenGenerated=<div class="success">Token generated! <code>The value will not be displayed </code>anymore, please save it.</div>
```
Actually this string is never used, so I removed it.

Wadeck commented on core/src/main/resources/jenkins/security/apitoken/ApiTokenPropertyConfiguration/config.jelly:
```diff
@@ -0,0 +1,39 @@
+<j:jelly xmlns:j="jelly:core" xmlns:f="/lib/form">
+    <f:section title="${%API Token}">
+        <f:entry field="tokenGenerationOnCreationEnabled">
+            <f:checkbox title="${%Enable generation of Legacy API Token on user creation}"/>
```
(...) inside inline i18n is considered as an expression by Jelly, I just take that opportunity to create a `config.messages` instead of escaping parenthesis

Wadeck commented on core/src/main/resources/jenkins/security/apitoken/ApiTokenPropertyConfiguration/config.jelly:
```diff
@@ -0,0 +1,39 @@
+            <f:checkbox title="${%Enable generation of Legacy API Token on user creation}"/>
+        </f:entry>
+        <f:entry field="creationOfLegacyTokenEnabled">
+            <f:checkbox title="${%Enable capability of user without Legacy API Token to create new one}"/>
```
(...) inside inline i18n is considered as an expression by Jelly, I just take that opportunity to create a `config.messages` instead of escaping parenthesis

jeffret-b commented on core/src/main/resources/jenkins/security/apitoken/LegacyApiTokenAdministrativeMonitor/manage.properties:
```diff
@@ -0,0 +1,33 @@
+title=Manage Legacy API Token usage
+contextMessage=The following users have a legacy API token. \
+  Because legacy tokens are stored in a recoverable format, we recommend migrating to the new API token system. \
+  Additionally, in previous version of Jenkins, API tokens were automatically created for every new user. \
```
Minor English language comment: It would read a little better if "version" -> "versions".

jeffret-b commented on core/src/main/resources/jenkins/security/apitoken/LegacyApiTokenAdministrativeMonitor/manage.properties:
```diff
@@ -0,0 +1,33 @@
+HasMoreRecentlyUsedToken_tooltip=A recently used token is one that was created using the new system \n\
+  and used more recently than the user's legacy token. \n\n\
+  Note: the creation date of legacy tokens which have never been used are reset every time Jenkins is restarted, \n\
+  which means that the date may be inaccurate.
```
Maybe it's just me, but these extra line breaks look weird on my screen. They disrupt the reading flow. Generally it's better to let the browser figure out where to wrap lines. Except where the line breaks are productive to the flow of the information. In other words, the line breaks before "Note:" are good and improve clarity.

Comment applies to both of these tooltips.

jeffret-b commented on core/src/main/resources/jenkins/security/apitoken/LegacyApiTokenAdministrativeMonitor/message.properties:
```diff
@@ -0,0 +1,3 @@
+warningMessage=There are users that are still using a legacy API token. \
```
It would be a little nicer to change "who" -> "that".

See [Grammar Girl](https://www.quickanddirtytips.com/education/grammar/who-versus-that)

Wadeck replied:
So I think it's `that` -> `who` that you meant ?

jeffret-b commented on core/src/main/resources/jenkins/security/ApiTokenProperty/help-tokenStore.html:
```diff
@@ -0,0 +1,12 @@
+    Some good practices for keeping your API tokens secure are:
+    <ul>
+        <li>Use a different token for each application so that if an application is compromised you can revoke its token individually.</li>
+        <li>Regenerate the tokens every 6 months (depending on your context). We put an indicator concerning the age of the token.</li>
```
Maybe "put" -> "display". It's a little hard to figure out where this indicator is being put.

jeffret-b commented on core/src/main/resources/jenkins/security/ApiTokenProperty/help-tokenStore.html:
```diff
@@ -0,0 +1,12 @@
+        <li>Regenerate the tokens every 6 months (depending on your context). We put an indicator concerning the age of the token.</li>
+        <li>Protect it like your password, as it allows other people to access Jenkins as you.</li>
+    </ul>
+</div>
```
Does this page also need a comment like "Note: the creation date of legacy tokens which have never been used are reset every time Jenkins is restarted, \n\
+  which means that the date may be inaccurate."

As on the "Manage Legacy API Token usage". Otherwise, it may confuse users who see inexplicable values here and don't see the admin page.

jeffret-b commented:
On the "Manage Legacy API Token usage" page, the value in "Fresh token?" and "Recent token?" columns kind of confuse me, when it shows warning.png. I think I kind of get what it means, but it took me a while of looking at it to get the idea.

jeffret-b reviewed (approved):
This is a great improvement!

I performed a pretty high-level review.

I reviewed the design and the various improvements through the conversation. All of that looks very good.

I played around with the basic interactions in the UI. All of that worked correctly and as I would expect. I left a few suggestions for wording or presentation that might could be improved.

I didn't review the code changes, the actual usage of the API key, or API key statistics.

Wadeck committed:
Address Jeff's rewording points
- add tooltip on the success/warning icon in manage page

Wadeck commented:
> when it shows `warning.png`

@jeffret-b I added a tooltip for the icons in that case, you're right, it was not that straightforward.

fcojfernandez reviewed (approved):
Great job @Wadeck 

It's been a quick review, but it really looks good to me.

varyvol commented on core/src/main/resources/jenkins/security/ApiTokenProperty/resources.js:
```diff
@@ -0,0 +1,128 @@
+                errorSpan.style.display = 'block';
+                errorSpan.innerHTML = json.message;
+
+                button.removeClassName('request-pending');
```
Shouldn't this be done if no error happens too?

Wadeck replied:
actually not because in the `else` there is `button.remove();`

varyvol commented on core/src/main/resources/jenkins/security/ApiTokenProperty/resources.js:
```diff
@@ -0,0 +1,128 @@
+                var tokenValue = json.data.tokenValue;
+                var tokenValueSpan = repeatedChunk.querySelector('.new-token-value');
+                tokenValueSpan.innerText = tokenValue;
+                tokenValueSpan.style.display = 'inline-block';
```
It feels weird seeing visualization aspects directly set on JS. Better adding a class or something like that and customizing the visualization by css?

Wadeck replied:
I also modify all the "style" attributes

varyvol commented on core/src/main/resources/jenkins/security/apitoken/LegacyApiTokenAdministrativeMonitor/resources.js:
```diff
@@ -0,0 +1,150 @@
+    for(var i = 0; i < concernedCheckBoxes.length ; i++){
+        var checkBox = concernedCheckBoxes[i];
+        if(!checkBox.checked){
+            mustCheck = true;
```
Breaking the loop here?

varyvol commented on core/src/main/resources/jenkins/security/apitoken/LegacyApiTokenAdministrativeMonitor/resources.js:
```diff
@@ -0,0 +1,150 @@
+        }
+    }
+    
+    for(var i = 0; i < allCheckBoxes.length ; i++){
```
Maybe forEach() to simplify code?

Wadeck replied:
I am not a huge fan of anonymous function and I do not want to introduce es6 arrows there. We gain 5 lines with your proposal, seems ok.

Wadeck replied:
But seems to make the tests to fail, so I revert

jeffret-b commented:
The changes look good. The tooltips read a lot better now. The tooltips on warning.png definitely help.

Wadeck committed:
Address Evaristo's comment on JS

Wadeck commented on core/src/main/java/hudson/model/User.java:
```diff
@@ -764,6 +764,10 @@ private boolean relatedTo(@Nonnull AbstractBuild<?,?> b) {
         return fullName;
     }
+    
+    public File getUserFolder(){
```
As it seems useful, I chose the Javadoc + @since approach.

Wadeck committed:
Address Daniel's comment

daniel-beck commented:
Test failures:

> `Cannot find function forEach in object [object StaticNodeList]`

Too modern JS?

Wadeck committed:
Undo the forEach change

daniel-beck commented:
Merge conflict.

daniel-beck commented:
Merge conflict.

Wadeck committed:
Merge branch 'master' into JENKINS-32442-32776_HASHED_TOKEN

daniel-beck commented:
Merge commit due to multiple contributors in the history.
